rust/protocol/src/sealed_sender.rs

```diff
@@ -879,8 +1019,214 @@ mod sealed_sender_v2 {
+/// 3. *Once per recipient:* Encrypt (1) using a shared secret derived from the private ephemeral
+///    key (2) and the recipient's public identity key.
+/// 4. *Once per recipient:* Add an authentication tag for (3) using a secret derived from the
+///    sender's private identity key and the recipient's public identity key.
+/// 5. Generate a symmetric key from (1) and use it to symmetrically encrypt the underlying
+///    [UnidentifiedSenderMessageContent] via [AEAD encryption]. *This step is only performed once
+///    per message, regardless of the number of recipients.*
+/// 6. Send the public ephemeral key (2), the encrypted random bytes (3), and the authentication tag
+///    (4) to the server, along with the encrypted message (5).
+///
+/// [AEAD encryption]: https://en.wikipedia.org/wiki/Authenticated_encryption#Authenticated_encryption_with_associated_data_(AEAD)
+///
+/// ## Pseudocode
+///```text
+/// ENCRYPT(message, R_i):
+///     M = Random(32)
+///     r = KDF(label_r, M, len=64)
+///     K = KDF(label_K, M, len=32)
+///     E = DeriveKeyPair(r)
+///     for i in num_recipients:
+///         C_i = KDF(label_DH, DH(E, R_i) || E.public || R_i.public, len=32) XOR M
+///         AT_i = KDF(label_DH_s, DH(S, R_i) || E.public || C_i || S.public || R_i.public, len=16)
+///     ciphertext = AEAD_Encrypt(K, message)
+///     return E.public, C_i, AT_i, ciphertext
+///
+/// DECRYPT(E.public, C, AT, ciphertext):
+///     M = KDF(label_DH, DH(E, R) || E.public || R.public, len=32) xor C
+///     r = KDF(label_r, M, len=64)
+///     K = KDF(label_K, M, len=32)
+///     E' = DeriveKeyPair(r)
+///     if E.public != E'.public:
+///         return DecryptionError
+///     message = AEAD_Decrypt(K, ciphertext) // includes S.public
+///     AT' = KDF(label_DH_s, DH(S, R) || E.public || C || S.public || R.public, len=16)
+///     if AT != AT':
+///         return DecryptionError
+///     return message
+///```
+///
+/// # Wire Format
+/// Multi-recipient sealed-sender does not use protobufs for its payload format. Instead, it uses
+/// a flat format marked with a [version byte](#the-version-byte). The format is different for
+/// [sending] and [receiving]. The decrypted content is
+/// a protobuf-encoded `UnidentifiedSenderMessage.Message` from `sealed_sender.proto`.
+///
+/// The public key used in Sealed Sender v2 is always a Curve25519 DJB key.
+///
+/// [sending]: #sent-messages
+/// [receiving]: #received-messages
+///
+/// ## The version byte
+///
+/// Sealed sender messages (v1 and v2) in serialized form begin with a version [byte][u8].
+/// This byte has the form:
+///
+/// ```text
+/// (requiredVersion << 4) | currentVersion
+/// ```
+///
+/// v1 messages thus have a version byte of `0x11`. v2 messages have a version byte
+/// of `0x22`. A hypothetical version byte `0x34` would indicate a message encoded
+/// as Sealed Sender v4, but decodable by any client that supports Sealed Sender v3.
+///
+/// ## Routing messages to recipients
```

Comment:
…but this doesn't…

Reply:
I have moved it into its own top-level heading above Wire Format!